test/core/circuit.spec.js

```diff
@@ -0,0 +1,172 @@
+/* eslint max-nested-callbacks: ["error", 8] */
+/* eslint-env mocha */
+'use strict'
+
+const chai = require('chai')
+const dirtyChai = require('dirty-chai')
+const expect = chai.expect
+const parallel = require('async/parallel')
+const series = require('async/series')
+const waterfall = require('async/waterfall')
+const API = require('ipfs-api')
+const IPFS = require('../../src')
+const bl = require('bl')
+const PeerInfo = require('peer-info')
+const PeerId = require('peer-id')
+const multiaddr = require('multiaddr')
+const isNode = require('detect-node')
+
+const crypto = require('crypto')
+
+const createTempRepo = require('../utils/create-repo-nodejs.js')
+
+chai.use(dirtyChai)
+
+function addAndCat (data, ipfsSrc, ipfsDst, callback) {
+  waterfall([
+    (cb) => ipfsDst.files.add(data, cb),
+    (res, cb) => ipfsSrc.files.cat(res[0].hash, cb),
+    (stream, cb) => stream.pipe(bl(cb))
+  ], callback)
+}
+
+function peerInfoFromObj (obj, callback) {
+  PeerInfo.create(PeerId.createFromB58String(obj.id), (err, peer) => {
+    if (err) {
+      return callback(err)
+    }
+
+    expect(err).to.not.exist()
+    obj.addresses.forEach((a) => peer.multiaddrs.add(multiaddr(a)))
+    callback(null, peer)
+  })
+}
+
+function createJsNode (addr) {
+  return new IPFS({
+    repo: createTempRepo(),
+    config: {
+      Addresses: {
+        Swarm: addr
+      },
+      Discovery: {
+        MDNS: {
+          Enabled: false
+        }
+      },
+      Bootstrap: [],
+      EXPERIMENTAL: {
+        Relay: {
+          Enabled: true,
+          HOP: {
+            Enabled: true,
+            Active: false
+          }
+        }
+      }
+    }
+  })
+}
```

Comment:
Why not use the default utils?

Reply:
Please avoid creating so much duplicated code for creating nodes (the same thing happened in libp2p). Always create code that is reusable and of low maintenance.